### PR TITLE
Remove C FILE PAL wrappers

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Remove the PAL wrappers around the C `FILE` APIs and update locations that were using the existing features.

The existing PAL layer provides the following features:

- Explicit "Text" mode:
   - Normalize line endings on read from `CRLF` to `LF`
   - Specify `'t'` as part of the mode parameter to `fopen`
   - This was used by ildasm for writing. Writing with text mode in the PAL does nothing (as the line endings are already normalized). Replaced with an `#ifdef` to only use explicit text mode on Windows.
- Implicit "Text" mode:
   - Normalize line endings on read from `CRLF` to `LF`
   - This was used across the product, but only used in two places that might read files with CRLF-style line endings.
      - Text-based PGO.
        - This format is a non-shipping format that is used for debugging/diagnostics purposes.
        - Text editors on Windows now support LF-based line endings, so I updated this to use the explicit non-text mode so files produced in the future can be read on all platforms (like the existing behavior).
      - ildasm's handling for showing source-line information
        - ildasm explicitly checks for CRLF line endings even when using the PAL, so we don't need the PAL to handle this.
- `EINTR` errno results
   - `EINTR` errors can only occur in the syscalls used by the C file APIs when a syscall is interrupted by a signal handler not marked with `SA_RESTART`. All signal handlers CoreCLR installs are marked with `SA_RESTART`.
- Clear `ferror` errors with `clearerr` before each operation.
   - I couldn't really validate that we don't depend on this behavior today, but given the minimal usage of these APIs in the runtime (mostly reading files in `/sys` and `/proc`), I'm not particularly worried.
   - Most file-manipulation scenarios cases in the CoreCLR code-base use the Win32 PAL (`CreateFile` and family) instead of the C PAL.

### CODE REVIEW

**Comment by janvorli** [comment]
> `EINTR` errors can only occur in the syscalls used by the C file APIs when a syscall is interrupted by a signal handler not marked with `SA_RESTART`. All signal handlers CoreCLR installs are marked with `SA_RESTART`.

How about signal handlers installed by user code? E.g. custom host or 3rd party libraries?

File: src/coreclr/pal/src/misc/cgroup.cpp [hunk 1]
```diff
@@ -179,7 +179,7 @@ class CGroup
         char *mountpath = nullptr;
         char *mountroot = nullptr;
 
-        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "rb");
```
Comment (janvorli) [inline src/coreclr/pal/src/misc/cgroup.cpp:182]: The stuff in this file doesn't need to be changed as it is always compiled against the system headers.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 2]
```diff
@@ -1170,7 +1170,7 @@ static size_t GetCurrentVirtualMemorySize()
     char* line = nullptr;
 
     // process virtual memory size is reported in the first column of the /proc/self/statm
-    FILE* file = fopen("/proc/self/statm", "r");
+    FILE* file = fopen("/proc/self/statm", "rb");
```
Comment (janvorli) [inline src/coreclr/gc/unix/gcenv.unix.cpp:1173]: This file doesn't need to be modified, it is compiled against system headers, not PAL.

File: src/coreclr/gc/unix/cgroup.cpp [hunk 3]
```diff
@@ -213,7 +213,7 @@ class CGroup
         char *mountpath = nullptr;
         char *mountroot = nullptr;
 
-        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "rb");
```
Comment (janvorli) [inline src/coreclr/gc/unix/cgroup.cpp:216]: This file doesn't need to be modified either, it is compiled against system headers, not PAL.

File: src/coreclr/pal/src/thread/process.cpp [hunk 4]
```diff
@@ -1690,7 +1690,7 @@ GetProcessIdDisambiguationKey(DWORD processId, UINT64 *disambiguationKey)
     snprintf(statFileName, sizeof(statFileName), "/proc/%d/stat", processId);
     _ASSERTE(chars > 0 && chars <= (int)sizeof(statFileName));
 
-    FILE *statFile = fopen(statFileName, "r");
+    FILE *statFile = fopen(statFileName, "rb");
```
Comment (janvorli) [inline src/coreclr/pal/src/thread/process.cpp:1693]: This file doesn't need to be modified, it is compiled against system headers.

**Comment by jkoritzinsky** [comment]
> > `EINTR` errors can only occur in the syscalls used by the C file APIs when a syscall is interrupted by a signal handler not marked with `SA_RESTART`. All signal handlers CoreCLR installs are marked with `SA_RESTART`.
> 
> How about signal handlers installed by user code? E.g. custom host or 3rd party libraries?

For signal handlers installed with the `PosixSignalRegistration` API, it looks like we're good. For custom hosts or 3rd party libraries that install signal handlers without SA_RESTART, we'd have a problem.

However, if the GC doesn't use the PAL headers (as you said in the later comments on some of the files), then the only places that use the `fopen` API are off-by-default logging scenarios, all of which other than `DOTNET_ReadyToRunLogFile` are Debug/Checked only, superpmi, or in non-CoreCLR processes like ilasm or ildasm. Everywhere else that does file manipulation uses the Win32 PAL, not the C APIs.

Therefore, I don't think we need to worry about `EINTR` error codes, especially as even the Mono runtime doesn't seem to check for them when using the C API instead of direct syscalls (they do check when directly calling `write` or `open`, but not `fwrite` or `fopen`).

**Comment by jkoritzinsky** [comment]
cc: @EgorBo for review of the text-based PGO format change (using LF line endings across all platforms instead of CRLF). We can also change this back and have it be LF on non-Windows and CRLF on Windows. I'd rather not support CRLF on non-Windows for this though.

**Review by EgorBo — APPROVED** [review]
Text PGO changes LGTM cc @AndyAyersMS @jakobbotsch hope it won't break any internal tooling for you

**Comment by lambdageek** [comment]
> especially as even the Mono runtime doesn't seem to check for them when using the C API instead of direct syscalls (they do check when directly calling write or open, but not `fwrite` or `fopen`).

Mono probably should check for `EINTR` in more places, to be honest - we're not a great role model for this.

(Although I'd guess we mostly use the `FILE*` apis in non-shipping debug scenarios and in the AOT compiler.  The AOT compiler doesn't load arbitrary 3rd party native code and I'm not even sure if we set up our own signal handlers in it.)

**Comment by jkoritzinsky** [comment]
@janvorli do you have any more feedback here?

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**